backend/src/main/kotlin/app/ehrenamtskarte/backend/verification/CanonicalJson.kt

```diff
@@ -0,0 +1,92 @@
+package app.ehrenamtskarte.backend.verification
+
+import com.google.protobuf.Descriptors
+import com.google.protobuf.Descriptors.FieldDescriptor.Type
+import com.google.protobuf.GeneratedMessageV3
+
+class CanonicalJson {
+    companion object {
+        /**
+         * Returns a canonical JSON object from the given protobuf message.
```

Comment:
This comment just shows how this function, does break the single responsibility principle. If we want to change this to cleaner code (which i really prefer), it would be good to:
1) create a function `assertMessageIsValid` 
2) create three functions, which each of the asserts one of the three listed asserations and has a meaningful name, so the name of the functions tells us the same information as the content of the corresponding comment bullet point
3) Create a function, which adds the entries to the object
4) Only call the two functions described in 1) and 3) in the `messageToMap` function.
5) Get rid of the comment, but the last line, as our code is now self explaining.

With this canges, we have code that is self explaining, we do not have the duplication of code and comment, which can lead to inconsistency in the future, we follow the single responsibility principle and the statements in every function are at the same level of abstraction.

Reply:
If it is easy to read, why dis there a comment explaining what is going on?
In my opinion the function solves two responsibilities: Asserting the input is valid and transforming it.
And there are different levels of abstraction in this function.
Function length should be no longer than 20 lines, according to uncle bob and his clean code book, which is quite the gold standard for clean code. This function is more than 40 lines.
For helping the user not to get lost in a lot of functions, we easily can use private functions.
(Just wanted to explain this is not a personal opinion thing on my site, but following the clean code principles of robert c. martin.)

I am a big fan of clean coding and highly recommend to follow these principles, but I am also fine if we as a team decide we do not want to follow them.

Reply:
Clearly, code style is always opinionated. "clean code" principles are not the gold standard in my opinion.

My personal opinion is that dogmatic use of "clean code" principles do not serve the purpose of good readability and maintainability. Clearly it makes sense to not have >100 loc functions (unless in cases where it actually makes sense :D ).
In this very specific case, the actual number of lines that *do* something is maybe 25.

> If it is easy to read, why dis there a comment explaining what is going on?

As this function is central in the stable hashing procedure, it needs to be documented well. Therefore, adding a doc-string for this very specific function that explains the contract of the function (such that it can be replicated in other languages) in English does make sense in my opinion (at least it made sense to me at time of writing). Also adding comments where and how the procedure may be extended to keep the stability property seems reasonable to me. This cannot be documented in code.

> In my opinion the function solves two responsibilities: Asserting the input is valid and transforming it.

We could try checking the preconditions in a separate function. My feeling is that it won't increase readability.

> And there are different levels of abstraction in this function.

What do you mean by "levels of abstraction"?

Reply:
Yup, definitely personal preference :D

Note, how the relevant code now is spread across ~75 lines and as a user of the public function I have to read private functions to understand the contract.

But let's close the discussion here^^ (unless you want to discuss it further :D )